### ISSUE TITLE
[UX] Append the local task text to the page <title> tag

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1508]
Say for example that you are editing the Account menu (```admin/structure/menu/manage/user-menu```). The page title is ```<title>Account menu | Backdrop Test Site</title>``` - OK.

Now switch to the "Configure" tab (```admin/structure/menu/manage/user-menu/configure```). Page title is still ```<title>Account menu | Backdrop Test Site</title>```. I think it should be ```<title>Account menu: Configure | Backdrop Test Site</title>``` instead.

Tagging "UX" because this text is shown on the browser tabs and if one has for example in this case both tabs open, they display the same title and so they have to click to see which one is which.

**Comment by klonos** [comment]
The PR handles 95% of the cases. The other 5% is to handled by follow-up issues like #1456.

I initially expected dozens of tests to fail (because of the title format changes), but then everything came back green. Than I remembered/realized that I am actually changing the head `<title>` and did not touch the page title, so there are no failing tests like in https://github.com/backdrop/backdrop/pull/1731

The approach I have taken is to add a new `menu_local_tasks_get_active_title()` function that behaves similarly to [`menu_local_tasks()`](https://api.backdropcms.org/api/backdrop/core!includes!menu.inc/function/menu_local_tasks/1) but returns the title of the active local primary/secondary tabs. I then use that + [`menu_get_active_title()`](https://api.backdropcms.org/api/backdrop/core!includes!menu.inc/function/menu_get_active_title/1) to construct various additional parts that comprise the head `<title>`.

Previously, we were using only page title, site name and site slogan. We were storing these parts in an array and were imploding that with `|` as separator. In cases where the site slogan was not specified in a site, the head `<title>` would be like so:

`Current page title | Site name`

...if a slogan was specified, you'd get something like this instead:

`Current page title | Site name | Site slogan`

...if I kept the same practice of adding parts to the array and then imploding everything still using `|` as separator, we'd end p with something like this:

`Current page title | Current primary tab | Current secondary tab | Site name | Site slogan`

I prefer either `:` (or `>`) for separating the parts prior to the site name/slogan, then `|` to separate these "path" parts from the name/slogan combo, and finally `,` (or `-`) to separate the site name and the slogan. So, my PR produces head `<title>` tags like so:

`Current page title: Current primary tab: Current secondary tab | Site name, Site slogan`

Feel free to bike-shade on the separators according to what you use and what you've seen being used in most sites and I'll adapt the PR accordingly once we settle on something.

Also feel free to play around with the PR sandbox and tell me if I missed any case where the `<title>` is not rendered as expected.

@jenlampton additionally I'd like you to please chime in on whether you feel this is a feature request, bug or task and suggest a milestone. When you get a chance 😉 

Talking about milestones: the added function is not an API change per se ...it's an API *addition*, so please don't say 2.x. Now, I have not touched the names of the existing variables within the `template_preprocess_page()` function (`$variables['head_title_array']` and `$variables['head_title'] ` specifically), but I did make the following changes to the key names of the `$head_title` array:

`$head_title['title']` changed to `$head_title['page_title']`
`$head_title['name']` changed to `$head_title['site_name']`
`$head_title['slogan']` changed to `$head_title['site_slogan']`
added `$head_title['primary_tab']`
added `$head_title['secondary_tab']`

If you think that this is an API change, I can always revert the renamed keys (perhaps with a not to change them in 2.x for better beginner DX).

**Comment by klonos** [comment]
...some examples of paths and their respective head `<title>` tags before/after

- `/admin/content`:
  before: `Content | Test site | with a slogan`
  after: `Content: Manage content | Test site, with a slogan`
- `/admin/content/comment`:
  before: `Content | Test site | with a slogan`
  after: `Content: Manage comments: Published comments | Test site, with a slogan`
- `/admin/content/comment/approval`:
  before: `Content | Test site | with a slogan`
  after: `Content: Manage comments: Unapproved comments | Test site, with a slogan`
- `/admin/structure/types`:
  before: `Content types | Test site | with a slogan`
  after: `Content types: List content types | Test site, with a slogan`
- `/admin/structure/types/manage/page/display`:
  before: `Page | Test site | with a slogan`
  after: `Page: Manage displays | Test site, with a slogan`
- `/admin/structure/menu/manage/main-menu`
  before: `Primary navigation | Test site | with a slogan`
  after: `Primary navigation: Edit links | Test site, with a slogan`
- `/admin/structure/menu/manage/main-menu/configure`:
  before: `Primary navigation | Test site | with a slogan`
  after: `Primary navigation: Configure | Test site, with a slogan`

Website: http://1734.backdrop.backdrop.qa.backdropcms.org
Username: admin
Password: xeAGBxYK

**Comment by Graham-72** [comment]
@klonos you may be interested to see this article about the origin of bike-shed discussions. https://en.wikipedia.org/wiki/Law_of_triviality I'm a great fan of Parkinson's Laws.

**Comment by klonos** [comment]
@Graham-72 I think I've read that wikipedia article in the (distant) past. Are you trying to imply that I am wasting my time/energy with trivial things while missing the "big game"? ...if so, then that is true 😛. I am testing things and trying to learn more of the internals of Drupal/Backdrop and as I try to improve my laughable php skills, I only find trivial/easy things to work on.

**Comment by Graham-72** [comment]
>Are you trying to imply that I am wasting my time/energy with trivial things

@klonos no. not at all. Apologies if it seemed I was criticising your valuable work. I was triggered by a mention of 'bike-shade' rather than 'bike-shed' (spell-checker issues?) so I thought I would track down the source of the expression. Completely off the point, sorry. 😧

**Comment by klonos** [comment]
No apologies required @Graham-72 even if your comment was criticism. I know you long enough to understand that it would be well-intended and in a "playful" manner. That's why there's a silly smiley on my comment there... to show that there was not offense on my part 😄

**Comment by Dinornis** [comment]
👍 
I use browser tabs almost excessively and find this very helpful.
Since this is only for admins I think the tags can be even shorter and do away with the slogans:

- `/admin/content`:
  before: `Content | Test site | with a slogan`
  after: `Manage content | Test site`
- `/admin/content/comment`:
  before: `Content | Test site | with a slogan`
  after: `Published comments | Test site`
- `/admin/content/comment/approval`:
  before: `Content | Test site | with a slogan`
  after: `Unapproved comments | Test site`
- `/admin/structure/types`:
  before: `Content types | Test site | with a slogan`
  after: List content types | Test site`
- `/admin/structure/types/manage/page/display`:
  before: `Page | Test site | with a slogan`
  after: `Page: Manage displays | Test site`
- `/admin/structure/menu/manage/main-menu`
  before: `Primary navigation | Test site | with a slogan`
  after: `Primary navigation: Edit links | Test site`
- `/admin/structure/menu/manage/main-menu/configure`:
  before: `Primary navigation | Test site | with a slogan`
  after: `Primary navigation: Configure | Test site`
- `/admin/config/development/configuration/single/export`
  before: `Configuration management | Test site | with a slogan`
  after: `Configuration management: Export | Test site`

**Comment by klonos** [comment]
@Dinornis I need to come back to this at some point when I have time. Most head titles in the admin UI work much better with my PR, but there are exceptions where things do not work as expected. Notably, when editing nodes, you get `%node_title%`: Edit | %site_name%` which is an improvement, but when viewing nodes, you get `%node_title%`: View | %site_name%` which is a side effect. I could adjust my current PR to be checking whether the current path is backend/admin or frontend and have different patterns for the head titles per case (not showing the slogan on admin as you suggested being a part of this). Another idea is to implement a hook that would allow overriding the default way the head title is generated, even making the overriding work granularly per head title part (page parent title/current page title/site name/site slogan/separators between parts). Will see.

Now, all I need is general feedback for whether people like this change or not, and thoughts on the code implementation.

**Comment by quicksketch** [comment]
> but I did make the following changes to the key names of the $head_title array:

I would recommend keeping the keys the same. It's a minor API change but one none-the-less.

Implementation overall looks pretty good.

I'd bike-shed-propose that perhaps we shouldn't include the site slogan in the head title at all, with the exception possibly the home page. Even that seems strange to me though.

> but when viewing nodes, you get %node_title%: View | %site_name% which is a side effect.

I'd probably say if you're on the *default* primary tab (such as "View"), we shouldn't be adding it to the head title, just leave it off. That will also prevent redundant headers like "Content: Manage content" or "URL Aliases: List URL Aliases".

**Comment by klonos** [comment]
OK, coming back to this after a really long time (sorry 😓). I have filed a new PR, because it seems that the original forked repo I had back in 2017 was deleted, and replaced with a new.

In this reincarnation, I have done the following:

> I would recommend keeping the keys the same. It's a minor API change but one none-the-less.

Agreed. I have kept the the array keys the same.

> I'd bike-shed-propose that perhaps we shouldn't include the site slogan in the head title at all

Hmm 🤔 ...[we already include the slogan](https://github.com/backdrop/backdrop/blob/1.x/core/includes/theme.inc#L2705) (if it is set), so that would be a regression for existing sites. We could implement that as a "Include site slogan in the head <title>" checkbox, in the site info form - but perhaps separate ticket (or contrib)?

> I'd probably say if you're on the *default* primary tab (such as "View"), we shouldn't be adding it to the head title, just leave it off. That will also prevent redundant headers like "Content: Manage content" or "URL Aliases: List URL Aliases".

Also hmm 🤔 ...that would not work in certain cases. The default primary tab for the Dashboard is "Overview", so in that case, having it be "Dashboard: Overview" provides better context (which is also consistent with its "Settings" tab, which is better as "Dashboard: Settings"). I don't see an easy way out of this, but still working on it. In the meantime, I would love 👀 for review/testing as well as some more feedback/ideas.

**Comment by klonos** [comment]
...so with the current implementation, instead of this:

![Screen Shot 2019-11-23 at 12 59 06 pm](https://user-images.githubusercontent.com/2423362/69471219-20fdba00-0df1-11ea-8daa-d880c0a1886c.png)

...you get this:

![Screen Shot 2019-11-23 at 1 00 31 pm](https://user-images.githubusercontent.com/2423362/69471235-42f73c80-0df1-11ea-93bc-4047be9d505b.png)

**Comment by klonos** [comment]
Problems that I still need to work on:

- `admin/appearance` is "Appearance: List themes" 👍 
- `admin/appearance/install` is "Install themes: Install new themes" 👎 ...but I believe that this is a problem with the manually set page title (via `backdrop_set_title()`)
- `admin/appearance/update` is "Appearance: Update themes" 👍 

---

- `admin/modules` is "Modules: List modules" 👍 (although could be changed to "Functionality: List module" for consistency)
- `admin/modules/install` is "Install modules: Install new modules" 👎 (same issue as with themes)
- `admin/modules/update` is "Functionality: Update modules" 👍 
- `admin/modules/uninstall` is "Functionality: Uninstall modules" 👍 

---

- `admin/structure/layouts` is "Layouts: List layouts" 👍
- `admin/structure/layouts/install` is "Install layouts: Install new layouts" 👎 (same issue as with themes/modules)
- `admin/structure/layouts/update` is "Layouts: Update layouts" 👍 
- `admin/structure/layouts/settings` is "Layouts: Templates" 👍 

---

Anything under `admin/structure` is "Whatever" 👎 ...I would like it to be "Structure: Whatever" instead

---

- `admin/structure/types/manage/page` is "Page: Configure" 👍 (should it be "Content types: Page: Manage displays" though?)
- `admin/structure/types/manage/page/fields` is "Page: Manage fields" 👍 
- `admin/structure/types/manage/page/display` is "Page: Manage displays" 👍 
- `admin/structure/types/manage/page/comment/fields` is "Page: Comment fields" 👍 
- `admin/structure/types/manage/page/display/default` is "Page: Default display: Manage displays" 👎 ...should ideally be "Page: Manage displays: Default display", or even shorter "Page: Manage Default display"
- `admin/structure/types/manage/page/comment/display/default` is " : Default display: Comment display" 👎 ...not introduced by this PR (but we could include the fix here)

...???

**Comment by klonos** [comment]
I am thinking that there may never be a perfect logic that will get everything 100% right all the time, or if we implement it, it will be very complex and hard to understand/maintain. So I would like to propose that we introduce a new `backdrop_set_head_title()` function for any edge cases, and when for UX reasons we need to have a different page title than the HTML head title?

**Comment by BWPanda** [comment]
I'm not sure about this... Tab titles don't seem to be the place to put the breadcrumb trail for where you are on the site... I agree that we should change the current situation, but I wonder if there's a more succinct way of doing it?

I also think there shouldn't be more than two sections in the title, separated by a '|'. So it should either be '[Title] | [Site name]', or '[Site name] | [Slogan]'.
